sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -931,6 +931,14 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression {
         $evPrim = $result.copy();
       """
   }
+
+  override def sql: Option[String] = {
+    if (foldable) {
+      Literal.create(eval(), dataType).sql
+    } else {
+      child.sql.map(childSQL => s"CAST($childSQL AS ${dataType.sql})")
```

Comment:
Yeah, I also found this change introduced a bug and caused test failure. Actually the major concern here is casting to complex types. HiveQL doesn't allow casting to complex types.